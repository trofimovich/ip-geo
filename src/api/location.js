import axios from 'axios';

const ipstackAPI = axios.create({
  baseURL: 'http://api.ipstack.com/',
  timeout: 1000,
});

const DEFAULT_REQUEST_DATA = 'check'; // used to receive location based on user IP

export const getLocationInfo = (requestData = DEFAULT_REQUEST_DATA) => {
  const fields = ['country_name', 'ip', 'location', 'latitude', 'longitude', 'city'];
  return ipstackAPI.get(`/${requestData}?fields=${fields.join(',')}&access_key=${process.env.IPSTACK_API_KEY}`);
};
