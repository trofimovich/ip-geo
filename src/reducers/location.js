import { combineReducers } from 'redux';

const isFetching = (state = false, action) => {
  switch (action.type) {
    case 'LOCATION_REQUEST_INIT':
      return true;
    case 'LOCATION_REQUEST_SUCCESS':
    case 'REQUEST_ERROR':
      return false;
    default:
      return state;
  }
};

const userLocation = (state = null, action) => {
  switch (action.type) {
    case 'LOCATION_USER_REQUEST_INIT':
      return state;
    case 'LOCATION_USER_REQUEST_SUCCESS':
      return action.location;
    case 'REQUEST_ERROR':
      return state;
    default:
      return state;
  }
};

const activeLocation = (state = null, action) => {
  switch (action.type) {
    case 'LOCATION_REQUEST_INIT':
      return state;
    case 'LOCATION_REQUEST_SUCCESS':
      return action.location;
    case 'REQUEST_ERROR':
      return state;
    default:
      return state;
  }
};

const searchHistory = (state = [], action) => {
  switch (action.type) {
    case 'LOCATION_REQUEST_SUCCESS':
      return [...state, action.location];
    default:
      return state;
  }
};

export default combineReducers({
  isFetching,
  userLocation,
  activeLocation,
  searchHistory,
});
