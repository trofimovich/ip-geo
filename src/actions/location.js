import * as api from '../api/location';

export const findUserLocation = () => (dispatch) => {
  dispatch({ type: 'LOCATION_USER_REQUEST_INIT' });
  return api
    .getLocationInfo()
    .then(({ data }) => dispatch({ type: 'LOCATION_USER_REQUEST_SUCCESS', location: data }))
    .catch((error) => dispatch({ type: 'LOCATION_USER_REQUEST_ERROR', error }));
};

export const findLocation = (formData) => (dispatch) => {
  dispatch({ type: 'LOCATION_REQUEST_INIT' });
  return api
    .getLocationInfo(formData)
    .then(({ data }) => dispatch({ type: 'LOCATION_REQUEST_SUCCESS', location: { ...data, searchString: formData } }))
    .catch((error) => dispatch({ type: 'LOCATION_REQUEST_ERROR', error }));
};
