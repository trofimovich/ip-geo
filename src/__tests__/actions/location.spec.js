import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import nock from 'nock';
import { findLocation } from '../../actions/location';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('actions', () => {
  let store;
  let fetchedLocation = {
    country_name: 'US'
  }

  beforeEach(() => {
    store = mockStore({});
  });

  afterEach(() => {
    nock.cleanAll();
  });

  it('creates "LOCATION_REQUEST_INIT", "LOCATION_REQUEST_ERROR" in case of error', () => {
    nock('https://api.ipstack.com')
      .get('/invalid.com?fields=country_name,ip,location,latitude,longitude,city&access_key=undefined')
      .reply(500, null, {'Access-Control-Allow-Origin': '*'});

    return store.dispatch(findLocation('invalid.com')).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it('creates "LOCATION_REQUEST_INIT", "LOCATION_REQUEST_SUCCESS" in case of error', () => {
    nock('https://api.ipstack.com')
      .get('/valid.com?fields=country_name,ip,location,latitude,longitude,city&access_key=undefined')
      .reply(200, fetchedLocation, {'Access-Control-Allow-Origin': '*'});

    return store.dispatch(findLocation('valid.com')).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
  })
});
