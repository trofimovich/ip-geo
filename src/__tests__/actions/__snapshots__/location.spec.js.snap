// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions creates "LOCATION_REQUEST_INIT", "LOCATION_REQUEST_ERROR" in case of error 1`] = `
Array [
  Object {
    "type": "LOCATION_REQUEST_INIT",
  },
  Object {
    "error": [Error: Request failed with status code 500],
    "type": "LOCATION_REQUEST_ERROR",
  },
]
`;

exports[`actions creates "LOCATION_REQUEST_INIT", "LOCATION_REQUEST_SUCCESS" in case of error 1`] = `
Array [
  Object {
    "type": "LOCATION_REQUEST_INIT",
  },
  Object {
    "location": Object {
      "country_name": "US",
      "searchString": "valid.com",
    },
    "type": "LOCATION_REQUEST_SUCCESS",
  },
]
`;
